core/services/pipeline/runner.go

```diff
@@ -137,12 +136,35 @@ func (m *memoryTaskRun) results() (a []Result) {
 	return
 }
 
+func (m *memoryTaskRun) numPredecessorsRemaining() int {
+	m.predMu.RLock()
+	defer m.predMu.RUnlock()
+	return m.predecessorsRemaining
+}
+
+func (m *memoryTaskRun) tryToClaim() (ok bool) {
```

Comment:
hm it's a bit non-obvious what claiming has to do with `m.finished` - maybe `tryFinish` or `trySwitchToFinished` ?